SWDEV-545950 - Add hipStreamCopyAttributes API Implementation

satyanveshd opened the pull request:
## Motivation

Adds a new API hipStreamCopyAttributes to HIP runtime.

<!-- Explain the purpose of this PR and the goals it aims to achieve. -->

Adds the header and definition for the API hipStreamCopyAttributes

## Technical Details

Currently, only the stream attribute hipSynchronizationPolicy can be set by the user, and hence only this attribute can be copied from one stream to other during runtime.

## Test Plan

Added a unit test covering the basic sanity for this API

<!-- Explain any relevant testing done to verify this PR. -->

Verified various scenarios by running the catch test locally 

## Test Result

PASS

## Submission Checklist

- [x] Look over the contributing guidelines at https://github.com/ROCm/ROCm/blob/develop/CONTRIBUTING.md#pull-requests.

satyanveshd committed:
SWDEV-545950 - Add hipStreamCopyAttributes API Implementation

satyanveshd committed:
Add unit test for hipStreamCopyAttributes API

satyanveshd committed:
Add ChangeLog and nvidia mapping for the API

satyanveshd committed:
Update rocprofiler-sdk with new HIP API details

itrowbri commented:
@satyanveshd, it looks like you're just changing the commented version of HIP_RUNTIME_API_TABLE_STEP_VERSION in hip_api_trace.hpp. Could you update the actual definition and then update the rocprofiler-sdk ABI code to work with this new step version? It also looks like this other HIP PR (https://github.com/ROCm/rocm-systems/pull/653) is added a HIP function without updating the step version. Could you work @agunashe to update HIP_RUNTIME_API_TABLE_STEP_VERSION correctly across both of these PRs?

jrmadsen commented on projects/clr/hipamd/src/hip_api_trace.cpp:
```diff
@@ -2084,13 +2086,14 @@ HIP_ENFORCE_ABI(HipDispatchTable, hipLibraryLoadFromFile_fn, 497);
 //
 //  HIP_ENFORCE_ABI_VERSIONING(<table>, 9) <- 8 + 1 = 9
-HIP_ENFORCE_ABI_VERSIONING(HipDispatchTable, 501)
+HIP_ENFORCE_ABI_VERSIONING(HipDispatchTable, 502)
```
This should never be changed without bumping the step version

satyanveshd replied:
As far as I know, this ABI VERSIONING needs to be incremented for every new HIP API added.  And the HIP_RUNTIME_API_TABLE_STEP_VERSION is incremented only once for every new release.  For 7.2, this version is incremented to 15 already with the hipLibrary* API changes in #975 and hence all the new APIs will be under VERSION 15 itself until the next release.

satyanveshd replied:
Updated the HIP_RUNTIME_API_TABLE_STEP_VERSION as discussed for this new API which has 2 stream arguments.

jrmadsen commented on projects/rocprofiler-sdk/source/include/rocprofiler-sdk/cxx/enum_string.hpp:
```diff
@@ -1012,7 +1013,7 @@ static_assert(ROCPROFILER_HIP_RUNTIME_API_ID_LAST == 477);
 #elif HIP_RUNTIME_API_TABLE_STEP_VERSION == 14
 static_assert(ROCPROFILER_HIP_RUNTIME_API_ID_LAST == 496);
 #elif HIP_RUNTIME_API_TABLE_STEP_VERSION == 15
-static_assert(ROCPROFILER_HIP_RUNTIME_API_ID_LAST == 501);
+static_assert(ROCPROFILER_HIP_RUNTIME_API_ID_LAST == 502);
 #else
```
```suggestion
#elif HIP_RUNTIME_API_TABLE_STEP_VERSION == 15
static_assert(ROCPROFILER_HIP_RUNTIME_API_ID_LAST == 501);
#elif HIP_RUNTIME_API_TABLE_STEP_VERSION == 16
static_assert(ROCPROFILER_HIP_RUNTIME_API_ID_LAST == 502);
#else
```

satyanveshd replied:
As far as I understand, The STEP_VERSION is incremented once for every new release.  For 7.2, it is already incremented to 15 with the earlier hipLibrary* APIs.  Please let me know if the understanding is incorrect.

satyanveshd replied:
Done

jrmadsen commented on projects/rocprofiler-sdk/source/include/rocprofiler-sdk/hip/api_args.h:
```diff
@@ -3323,6 +3323,11 @@ typedef union rocprofiler_hip_api_args_t
+        hipStream_t dst;
+        hipStream_t src;
+    } hipStreamCopyAttributes;
 #endif
```
```suggestion
#endif
#if HIP_RUNTIME_API_TABLE_STEP_VERSION >= 16
    struct
    {
        hipStream_t dst;
        hipStream_t src;
    } hipStreamCopyAttributes;
#endif
```

satyanveshd replied:
Done.

jrmadsen commented on projects/rocprofiler-sdk/source/include/rocprofiler-sdk/hip/runtime_api_id.h:
```diff
@@ -558,6 +558,7 @@ typedef enum rocprofiler_hip_runtime_api_id_t  // NOLINT(performance-enum-size)
     ROCPROFILER_HIP_RUNTIME_API_ID_hipLibraryUnload,
     ROCPROFILER_HIP_RUNTIME_API_ID_hipLibraryGetKernel,
     ROCPROFILER_HIP_RUNTIME_API_ID_hipLibraryGetKernelCount,
+    ROCPROFILER_HIP_RUNTIME_API_ID_hipStreamCopyAttributes,
```
```suggestion
#endif
#if HIP_RUNTIME_API_TABLE_STEP_VERSION >= 16
    ROCPROFILER_HIP_RUNTIME_API_ID_hipStreamCopyAttributes,
```

satyanveshd replied:
Done.

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/hip/abi.cpp:
```diff
@@ -599,6 +599,7 @@ ROCP_SDK_ENFORCE_ABI(::HipDispatchTable, hipLibraryLoadFromFile_fn, 497);
 ROCP_SDK_ENFORCE_ABI(::HipDispatchTable, hipLibraryUnload_fn, 498);
 ROCP_SDK_ENFORCE_ABI(::HipDispatchTable, hipLibraryGetKernel_fn, 499);
 ROCP_SDK_ENFORCE_ABI(::HipDispatchTable, hipLibraryGetKernelCount_fn, 500);
+ROCP_SDK_ENFORCE_ABI(::HipDispatchTable, hipStreamCopyAttributes_fn, 501);
```
```suggestion
#endif
#if HIP_RUNTIME_API_TABLE_STEP_VERSION >= 16
ROCP_SDK_ENFORCE_ABI(::HipDispatchTable, hipStreamCopyAttributes_fn, 501);
```

satyanveshd replied:
Done

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/hip/abi.cpp:
```diff
@@ -632,7 +633,7 @@ ROCP_SDK_ENFORCE_ABI_VERSIONING(::HipDispatchTable, 477)
 ROCP_SDK_ENFORCE_ABI_VERSIONING(::HipDispatchTable, 496)
 #elif HIP_RUNTIME_API_TABLE_STEP_VERSION == 15
-ROCP_SDK_ENFORCE_ABI_VERSIONING(::HipDispatchTable, 501)
+ROCP_SDK_ENFORCE_ABI_VERSIONING(::HipDispatchTable, 502)
```
```suggestion
ROCP_SDK_ENFORCE_ABI_VERSIONING(::HipDispatchTable, 501)
#endif
#if HIP_RUNTIME_API_TABLE_STEP_VERSION >= 16
ROCP_SDK_ENFORCE_ABI_VERSIONING(::HipDispatchTable, 502)
```

satyanveshd replied:
Done

jrmadsen commented on projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/hip/hip.def.cpp:
```diff
@@ -626,6 +626,7 @@ HIP_API_INFO_DEFINITION_V(ROCPROFILER_HIP_TABLE_ID_Runtime, ROCPROFILER_HIP_RUNT
 HIP_API_INFO_DEFINITION_V(ROCPROFILER_HIP_TABLE_ID_Runtime, ROCPROFILER_HIP_RUNTIME_API_ID_hipLibraryUnload, hipLibraryUnload, hipLibraryUnload_fn, library);
 HIP_API_INFO_DEFINITION_V(ROCPROFILER_HIP_TABLE_ID_Runtime, ROCPROFILER_HIP_RUNTIME_API_ID_hipLibraryGetKernel, hipLibraryGetKernel, hipLibraryGetKernel_fn, pKernel, library, name);
 HIP_API_INFO_DEFINITION_V(ROCPROFILER_HIP_TABLE_ID_Runtime, ROCPROFILER_HIP_RUNTIME_API_ID_hipLibraryGetKernelCount, hipLibraryGetKernelCount, hipLibraryGetKernelCount_fn, count, library);
+HIP_API_INFO_DEFINITION_V(ROCPROFILER_HIP_TABLE_ID_Runtime, ROCPROFILER_HIP_RUNTIME_API_ID_hipStreamCopyAttributes, hipStreamCopyAttributes, hipStreamCopyAttributes_fn, dst, src);
```
```suggestion
#endif
#if HIP_RUNTIME_API_TABLE_STEP_VERSION >= 16
HIP_API_INFO_DEFINITION_V(ROCPROFILER_HIP_TABLE_ID_Runtime, ROCPROFILER_HIP_RUNTIME_API_ID_hipStreamCopyAttributes, hipStreamCopyAttributes, hipStreamCopyAttributes_fn, dst, src);
```

satyanveshd replied:
Done

satyanveshd commented:
> @satyanveshd, it looks like you're just changing the commented version of HIP_RUNTIME_API_TABLE_STEP_VERSION in hip_api_trace.hpp. Could you update the actual definition and then update the rocprofiler-sdk ABI code to work with this new step version? It also looks like this other HIP PR (#653) is added a HIP function without updating the step version. Could you work @agunashe to update HIP_RUNTIME_API_TABLE_STEP_VERSION correctly across both of these PRs?

@itrowbri @jrmadsen As far as I understand, this HIP_RUNTIME_API_TABLE_STEP_VERSION is incremented only once for every new release.  For release 7.2, this version is incremented to 15 earlier in #975 and hence we added the new APIs under STEP_VERSION 15 itself. Also, I updated the rocprofiler-sdk ABI code accordingly.  But the actual issue seem to be different from the PSDB logs.

```
/longer_pathname_so_that_rpms_can_support_packaging_the_debug_info_for_all_os_profiles/src/rocm-systems/projects/rocprofiler-sdk/source/lib/rocprofiler-sdk/hip/stream.cpp:455:38: error: static assertion failed: function has more than one stream argument
  455 |             static_assert(stream_idx == (num_args - rstream_idx - 1),
  
```
  It seems the rocprofiler-sdk code doesn't expect two stream arguments in the HIP API. But this particular API hipStreamCopyAttributes have two arguments one for source stream and another for destination stream. And hence it fails. Please let me know if this understanding is right.

jrmadsen committed:
[rocprofiler-sdk] handle hipStreamCopyAttributes in stream tracing service

- this new HIP function has multiple stream arguments and needs to be skipped because it does not have an explicit create/destroy/set functionality

satyanveshd committed:
Update HIP_RUNTIME_API_TABLE_STEP_VERSION in clr and rocprofiler-sdk

satyanveshd committed:
Resolve merge conflicts

satyanveshd committed:
Merge branch 'develop' into amd/dev/sdittaka/SWDEV-545950

satyanveshd committed:
Merge branch 'develop' into amd/dev/sdittaka/SWDEV-545950
